File: app/src/main/java/com/unbounds/trakt/progress/Adapter.java
package com.unbounds.trakt.progress;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.squareup.picasso.Picasso;
import com.unbounds.trakt.ApiWrapper;
import com.unbounds.trakt.R;
import com.unbounds.trakt.api.model.Episode;
import com.unbounds.trakt.api.model.Show;
import com.unbounds.trakt.api.model.request.WatchedItems;
import com.unbounds.trakt.api.model.response.AddHistory;
import com.unbounds.trakt.api.model.response.WatchedProgress;

import java.util.ArrayList;
import java.util.List;

import rx.functions.Action1;

/**
 * Created by maclir on 2015-11-17.
 */
class Adapter extends RecyclerView.Adapter<Adapter.ViewHolder> {
    private final List<WatchedProgressWrapper> mData = new ArrayList<>();
    private final Context mContext;

    Adapter(final Context context) {
        mContext = context;
    }

    @Override
    public Adapter.ViewHolder onCreateViewHolder(final ViewGroup parent, final int viewType) {
        final View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.progress_item, parent, false);
        return new ViewHolder(view, new ViewHolder.OnClicked() {
            @Override
            public void onCheckClicked(final int position) {
                final WatchedProgressWrapper watchedProgressWrapper = mData.get(position);
                watchedProgressWrapper.setSelected(true);
                final Episode episode = watchedProgressWrapper.getWatchedProgress().getNextEpisode();
                ApiWrapper.postWatchedItems(new WatchedItems.Builder().addEpisode(episode).create()).subscribe(new Action1<AddHistory>() {
                    @Override
                    public void call(final AddHistory addHistory) {
                        //TODO check if added episode exists in not_added
                        ApiWrapper.getWatchedProgress(watchedProgressWrapper.getShow().getIds().getTrakt()).subscribe(new Action1<WatchedProgress>() {
                            @Override
                            public void call(final WatchedProgress watchedProgress) {
                                watchedProgressWrapper.setSelected(false);
                                if (watchedProgress.isCompleted()) {
                                    mData.remove(position);
                                    notifyItemRemoved(position);
                                } else {
                                    mData.set(position, watchedProgressWrapper.setWatchedProgress(watchedProgress));
                                    notifyItemChanged(position);
                                }
                            }
                        });
                    }
                });
            }
        });
    }

    @Override
    public void onBindViewHolder(final ViewHolder holder, final int position) {
        final WatchedProgressWrapper watchedProgressWrapper = mData.get(position);
        final WatchedProgress watchedProgress = watchedProgressWrapper.getWatchedProgress();
        final Show show = watchedProgressWrapper.getShow();
        final boolean selected = watchedProgressWrapper.isSelected();

        holder.mShowTitle.setText(show.getTitle());
        holder.mEpisodeTitle.setText(watchedProgress.getNextEpisode().getTitle());
        holder.mEpisodeNumber.setText(String.format("S%02dE%02d", watchedProgress.getNextEpisode().getSeason(), watchedProgress.getNextEpisode().getNumber()));
        Picasso.with(mContext).load(show.getImages().getPoster().getThumb()).into(holder.mShowPoster);
        holder.mCheck.setSelected(selected);
        holder.position = position;
    }

    @Override
    public int getItemCount() {
        return mData.size();
    }

    public void add(final WatchedProgressWrapper watchedProgressWrapper) {
        final int position = mData.size();
        mData.add(watchedProgressWrapper);
        notifyItemInserted(position);
    }

    public void clear() {
        mData.clear();
        notifyDataSetChanged();
    }

    static class ViewHolder extends RecyclerView.ViewHolder {
        public interface OnClicked {
            void onCheckClicked(final int position);
        }

        private final TextView mShowTitle;
        private final TextView mEpisodeTitle;
        private final TextView mEpisodeNumber;
        private final ImageView mShowPoster;
        private final View mCheck;
        private int position;

        public ViewHolder(final View view, final OnClicked listener) {
            super(view);
            mShowPoster = (ImageView) view.findViewById(R.id.progress_item_show_poster);
            mShowTitle = (TextView) view.findViewById(R.id.progress_item_show_title);
            mEpisodeTitle = (TextView) view.findViewById(R.id.progress_item_episode_title);
            mEpisodeNumber = (TextView) view.findViewById(R.id.progress_item_episode_number);
            mCheck = view.findViewById(R.id.progress_item_check);
            mCheck.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(final View view) {
                    if (mCheck.isSelected()) return;
                    mCheck.setSelected(true);
                    listener.onCheckClicked(position);
                }
            });
        }
    }
}

File: app/src/main/java/com/unbounds/trakt/view/progress/Adapter.kt
package com.unbounds.trakt.view.progress

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.LayoutRes
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.squareup.picasso.Picasso
import com.unbounds.trakt.R
import com.unbounds.trakt.viewmodel.NextEpisode
import kotlinx.android.synthetic.main.progress_item.view.*

/**
 * Created by maclir on 2015-11-17.
 */
class Adapter(private val listener: OnClicked) : ListAdapter<NextEpisode, Adapter.ViewHolder>(object : DiffUtil.ItemCallback<NextEpisode>() {
    override fun areItemsTheSame(oldItem: NextEpisode, newItem: NextEpisode) = oldItem.showId == newItem.showId
    override fun areContentsTheSame(oldItem: NextEpisode, newItem: NextEpisode) = oldItem == newItem
}) {
    interface OnClicked {
        fun onCheckClicked(episode: NextEpisode)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder = ViewHolder(parent.inflate(R.layout.progress_item))

    override fun onBindViewHolder(holder: ViewHolder, position: Int) = holder.bind(getItem(position))

    inner class ViewHolder(private val view: View) : RecyclerView.ViewHolder(view) {

        private var nextEpisode: NextEpisode? = null

        init {
            view.progress_item_check.setOnClickListener(View.OnClickListener {
                val episode = nextEpisode
                if (view.progress_item_check.isSelected || episode == null) return@OnClickListener
                view.progress_item_check.isSelected = true

                listener.onCheckClicked(episode)
            })
        }

        fun bind(episode: NextEpisode) {
            this.nextEpisode = episode

            with(episode) {
                Picasso.get().load(imageUrl).fit().into(view.progress_item_show_poster)
                view.progress_item_show_title.text = showTitle
                view.progress_item_episode_title.text = episodeTitle
                view.progress_item_progress_text.text = progressText
                view.progress_item_progress_bar.progress = progress
                view.progress_item_check.isSelected = selected
            }
        }
    }
}

fun ViewGroup.inflate(@LayoutRes layoutRes: Int, attachToRoot: Boolean = false): View {
    return LayoutInflater.from(context).inflate(layoutRes, this, attachToRoot)
}
